Fixing how subqueries use keys when building

patrickconroy opened the pull request:
Refs #6781.

Commit:
Fixing how subqueries use keys when building

ADmad commented:
Can you provide a test case please.

Commit:
Adding relevant test for deep hasMany w/subquery

patrickconroy commented:
TIL postgres sorts data weirdly.. fixing that now.

Commit:
Cleaning up test, but it'll probably still fail in postgres

Commit:
Testing the important thing here, nothing fancy

patrickconroy commented:
The fail looks unrelated now. I don't know how I'd fix this..

``
Fatal error: Uncaught exception 'Cake\Core\Exception\Exception' with message 'Unable to insert fixtures for "Cake\Test\TestCase\ORM\QueryRegressionTest" test case. SQLSTATE[23000]: [Microsoft][SQL Server Native Client 11.0][SQL Server]Cannot insert explicit value for identity column in table 'featured_tags' when IDENTITY_INSERT is set to OFF.' in C:\projects\cakephp\src\TestSuite\Fixture\FixtureManager.php:254
Stack trace:
``

lorenzo commented:
@patrickconroy it means that you are trying to insert a record having a known primary key, instead of letting the database create the next auto incremental id for you.

lorenzo commented on tests/Fixture/TagsTranslationsFixture.php:
```diff
@@ -0,0 +1,48 @@
+     * @var array
+     */
+    public $fields = [
+        'id' => ['type' => 'integer', 'null' => false],
```
I *think* you will need to add `'autoIncrement' => false` here, although I'm not sure

patrickconroy commented:
Hm, I'm not doing that. The fixture has a `records` array. There's no auto incremental id since the PK is id+locale. How/why would it attempt to increment a string?

ADmad commented on tests/Fixture/TagsTranslationsFixture.php:
```diff
@@ -0,0 +1,48 @@
+     * @var array
+     */
+    public $records = [
+        ['id' => '1', 'locale' => 'en_us', 'name' => 'tag 1 translated into en_us'],
```
Remove id from records and the postgres issue should be solved.

patrickconroy replied:
How would finding these make sense, then? The index in that array of those records !== the id..

ADmad replied:
The value for `id` field will be auto inserted by db. I don't think the `'autoIncrement' => false` @lorenzo suggested is required.

lorenzo replied:
He has duplicated `'id'` values @ADmad

patrickconroy replied:
The `id` value can't be auto-inserted, it doesn't make sense that way. There could be multiple `id=1`s or there could be 300 `id=17`s, it's a composite PK based on the id + locale. Maybe I'm missing something really simple about what you're saying though..

ADmad replied:
Error says: "**Cannot insert explicit value for identity column** in table 'featured_tags' when IDENTITY_INSERT is set to OFF".  Specifying the 'id' is trying to insert explicit value isn't it?

That said I have no experience with postgres, but have encountered similar error in the past (iirc) and removing ids from records fixed it.

ADmad replied:
Apologies, please ignore me. Like I said my suggestion was based off something i remembered to have encountered and not actual postgres experience :)

ADmad replied:
And I am an idiot too, the error is for ms sql not postgres.

patrickconroy replied:
Well, it looks like an ms sql error and not postgres, right? Are there other fixtures that use a non-id, non-autoincrement PK?

patrickconroy replied:
Hah no worries. The postgres stuff I got around another way; apparently the default sort on the records with a composite PK are opposite there.

Commit:
Attempting to fix fixtures in ms sql

Commit:
Just trying to get this to work..

patrickconroy commented:
So at this point I'm not sure how to fix that. Does anyone have any ideas? Is this an issue with how fixtures in ms sql work?

markstory commented on tests/Fixture/FeaturedTagsFixture.php:
```diff
@@ -0,0 +1,47 @@
+use Cake\TestSuite\Fixture\TestFixture;
+
+/**
+ * Class TagFixture
```
Lies :smile:

patrickconroy replied:
My copy/paste game wasn't great today I guess.

Commit:
Fixing some comments

ADmad commented on tests/Fixture/FeaturedTagsFixture.php:
```diff
@@ -0,0 +1,47 @@
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://book.cakephp.org/2.0/en/development/testing.html CakePHP(tm) Tests
+ * @since         1.2.0
```
Lies

patrickconroy replied:
Should I make that 3.0.7 ? I obviously don't know if this will make it to that, though.

lorenzo commented:
@markstory Do you happen to know how to make the fixture not create an auto increment id?

Commit:
Trying again to fix mssql

Commit:
Trying this?

patrickconroy commented:
I changed the fixture to not have set primary keys in the records. It seems weird that I had to.. but it looks like that fixed it. The important thing to know is that the test I added would fail if I didn't update `SelectableAssociationTrait::_buildSubquery`.

markstory commented:
@lorenzo you can set `autoIncrement = false`. If you need/want an autoincrement you will need to omit the autoIncrement columns from the fixture to make sqlserver happy.

markstory commented:
@patrickconroy looks good. Could you squash out the 'try this' commits?

lorenzo commented:
Squashed and merged in master. Thanks @patrickconroy !

patrickconroy commented:
Thanks for doing that! I'm afk until tomorrow.
